engine/engine.go

```diff
@@ -156,8 +157,11 @@ func (eng *Engine) Shutdown() {
 	eng.l.Lock()
 	if eng.shutdown {
 		eng.l.Unlock()
+		eng.shutdownLock.Lock()
+		defer eng.shutdownLock.Unlock()
```

Comment:
Trying to get my head around this one! :) your new lock is locked, and then immediately unlocked with the return on 162?